src/Avalonia.Controls/Presenters/TextPresenter.cs

```diff
@@ -416,29 +417,11 @@ public sealed override void Render(DrawingContext context)
                 return;
             }
 
-            var caretBrush = CaretBrush?.ToImmutable();
+            _caretPen ??= new(GetCaretBrush(CaretBrush, Background));
 
-            if (caretBrush is null)
-            {
-                var backgroundColor = (Background as ISolidColorBrush)?.Color;
+            var points = GetCaretPoints();
 
-                if (backgroundColor.HasValue)
-                {
-                    var red = (byte)~(backgroundColor.Value.R);
-                    var green = (byte)~(backgroundColor.Value.G);
-                    var blue = (byte)~(backgroundColor.Value.B);
-
-                    caretBrush = new ImmutableSolidColorBrush(Color.FromRgb(red, green, blue));
-                }
-                else
-                {
-                    caretBrush = Brushes.Black;
-                }
-            }
-
-            var (p1, p2) = GetCaretPoints();
-
-            context.DrawLine(new ImmutablePen(caretBrush), p1, p2);
+            context.DrawLine(_caretPen, points.Item1, points.Item2);
```

Comment:
Both old and a new code have an issue with mutable Background, which Color can be changed independently from the whole TextPresenter property. Some themes like Material already do this for their theming, i.e. change colors without changing replacing brushes. Which is a valid practice.